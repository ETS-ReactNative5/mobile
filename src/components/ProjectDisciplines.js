import React from 'react'
import {
  AlertIOS,
  Platform,
  PushNotificationIOS,
  ScrollView,
  StyleSheet,
  View
} from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'
import StyledText from './StyledText'
import { addIndex, filter, map, propEq } from 'ramda'
import { connect } from 'react-redux'
import {GLOBALS} from '../constants/globals'
import GoogleAnalytics from 'react-native-google-analytics-bridge'
import PropTypes from 'prop-types';
import Discipline from './Discipline'
import OverlaySpinner from './OverlaySpinner'
import NavBar from '../components/NavBar'
import { setState } from '../actions/index'
import { syncUserStore, setPushPrompted } from '../actions/user'

GoogleAnalytics.setTrackerId(GLOBALS.GOOGLE_ANALYTICS_TRACKING)
GoogleAnalytics.trackEvent('view', 'Home')

const topPadding = (Platform.OS === 'ios') ? 10 : 0

const mapStateToProps = (state) => ({
  user: state.user,
  isGuestUser: state.user.isGuestUser,
  isConnected: state.main.isConnected,
  isFetching: state.main.isFetching,
  pushPrompted: state.user.pushPrompted
})

const mapDispatchToProps = (dispatch) => ({
  setSelectedProjectTag(tag) {
    dispatch(setState('selectedProjectTag', tag))
  },
  setPushPrompted(value) {
    dispatch(setPushPrompted(value))
    dispatch(syncUserStore())
  },
})

export class ProjectDisciplines extends React.Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    if (this.shouldPromptForPermissions()) {
      setTimeout(()=> {
        this.promptRequestPermissions()
      }, 500)
    }
  }

  shouldPromptForPermissions() {
    return ((Platform.OS === 'ios') && (!this.props.user.pushPrompted))
  }

  promptRequestPermissions = () => {
    PushNotificationIOS.checkPermissions((permissions) => {
      if (permissions.alert === 0){
        AlertIOS.alert(
          'Allow Notifications?',
          'Zooniverse would like to occasionally send you info about new projects or projects needing help.',
          [
            {text: 'Not Now', onPress: () => this.requestIOSPermissions(false)},
            {text: 'OK', onPress: () => this.requestIOSPermissions(true)},
          ]
        )
      }
    })
  }

  requestIOSPermissions(accepted) {
    if (accepted) {
      PushNotificationIOS.requestPermissions()
    }
    this.props.setPushPrompted(true)
  }

  static renderNavigationBar() {
    return <NavBar showAvatar={true} />;
  }

  render() {
    const totalClassifications = this.props.user.totalClassifications
    const renderDiscipline = ({value, label, color}, idx) => {
      return (
        <Discipline
          icon={value}
          title={label}
          tag={value}
          key={idx}
          color={color}
          setSelectedProjectTag={() => {this.props.setSelectedProjectTag(value)}} /> )
    }

    const recent =
      <Discipline
        faIcon={'undo'}
        title={'Recent'}
        tag={'recent'}
        color={'rgba(0, 151, 157, 1)'}
        setSelectedProjectTag={() => {this.props.setSelectedProjectTag('recent')}} />

    const DisciplineList =
      <ScrollView>
        { !this.props.isGuestUser ? recent : null }
        {addIndex(map)(
          (discipline, idx) => { return renderDiscipline(discipline, idx) },
          filter(propEq('display', true), GLOBALS.DISCIPLINES)
        )}
      </ScrollView>

    const pluralizeClassification = ( totalClassifications > 1 ? 's' : '' )
    const totalClassificationsDisiplay =
      <StyledText
        additionalStyles={[styles.totalClassifications]}
        text={`${totalClassifications} total classification${pluralizeClassification}`} />

    return (
      <View style={styles.container}>
        <View style={styles.subNavContainer}>
            <StyledText additionalStyles={[styles.userName]}
              text = { this.props.isGuestUser ? 'Guest User' : this.props.user.display_name } />
            { totalClassifications > 0 ? totalClassificationsDisiplay : null }
        </View>
        <View style={styles.innerContainer}>
          { DisciplineList }
        </View>
        { this.props.isFetching ? <OverlaySpinner /> : null }
      </View>
    );
  }
}

const styles = EStyleSheet.create({
  container: {
    flex: 1,
  },
  subNavContainer: {
    borderBottomColor: '$lightGrey',
    borderBottomWidth: StyleSheet.hairlineWidth,
    paddingTop: 64 + topPadding,
    alignItems: 'center',
    justifyContent: 'flex-start',
    height: 112 + topPadding
  },
  userName: {
    color: '$darkTextColor',
    fontWeight: 'bold'
  },
  totalClassifications: {
    color: '$darkTextColor',
  },
  signOut: {
    backgroundColor: '$transparent',
  },
  signOutText: {
    color: '$darkTextColor',
    fontSize: 11,
  },
  messageContainer: {
    padding: 15,
  },
  innerContainer: {
    flex: 1,
    marginTop: 10,
  }
});

ProjectDisciplines.propTypes = {
  user: PropTypes.object,
  isGuestUser: PropTypes.bool,
  isFetching: PropTypes.bool,
  pushPrompted: PropTypes.bool,
  setSelectedProjectTag: PropTypes.func,
  setPushPrompted: PropTypes.func,
}

export default connect(mapStateToProps, mapDispatchToProps)(ProjectDisciplines)
