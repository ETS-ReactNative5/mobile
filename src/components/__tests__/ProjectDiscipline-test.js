import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { ProjectDisciplines } from '../ProjectDisciplines'

jest.useFakeTimers();

const projectActions = {
  fetchProjectsWithTags: () => {}
}

it('renders correctly', () => {
  const user ={ display_name: 'Fake User' }
  const tree = renderer.create(
    <ProjectDisciplines user={user} projectList={[]} projectActions={projectActions} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
