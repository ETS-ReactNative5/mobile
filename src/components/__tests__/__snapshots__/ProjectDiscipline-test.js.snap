// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={undefined}
>
  <View
    style={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": "Karla",
          },
          undefined,
        ]
      }
    >
      Fake User
    </Text>
  </View>
  <RCTScrollView
    contentContainerStyle={undefined}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={true}
      />
    }
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;
